### ISSUE TITLE
Implement multiple theming options

### ISSUE THREAD

**Issue by Hona** [ant-design-blazor/ant-design-blazor#1547]
Currently the only 'easy' way for users' to theme with is the replace the provided css files with precompiled ones, such as through an Ant Design CDN provider ([example](https://cdnjs.com/libraries/antd)). 

Before we go ahead with an implementation, lets discuss and agree on the best solution.

From the various discussions between the existing issues: (#31  #1240 #1540 #1043), my thoughts for the best solution is to implement all three of these solutions (to cover use cases):

## First Solution
**Use Case**: Static branding, for example a company just wants to change colors to match their brand, no user customisation - known at compile time.
**Proposed Solution**: CSS file middleware, with a class providing the theme, which compiles the file at server startup. This solution would only have a performance hit when starting up (which from testing less.js compilation in the browser was 5 seconds). For Blazor Server this would provide a very smooth experience for end users, not requiring modified CSS links, instead the server handles intercepting the request and serving the file. For Blazor WASM, we could implement some sort of local storage for the CSS file, so that it only has to compile the first time?

## Second Solution
**Use Case**: Same use case as the first solution, however the users should be able to modify their own themes. Their modified variables could be stored in a DB, and compiled at runtime/at request or when changed, and stored to disk matching a user ID. 
**Proposed Solution**: This would be an extension on the First Solution, whereby the middleware and service would still be added to the pipeline, but with an added DB/identifier, which then provides any overridden CSS files to users with configured themes, defaulting to the static theme (as defined in class as a default)

## Third Solution
**Use Case**: For WASM, or if the Blazor Server doesn't want to provide the compute resources/storage for all user's themes.
**Proposed Solution**: Less.js dynamic compilation in the users browser. The only notable issue is that this will take about ~5 seconds to compile for the user, which often times might cause them to leave the website. This could mix again with the First Solution, where the brands default theme is the baseline, so the first load is instant. We would have to provide some interface so that library user's don't have to use the JavaScript unless absolutely needed.

**Comment by Hona** [comment]
@ElderJames @anddrzejb @Epictek 

Please give your thoughts, lets discuss this so we are all on the same page for the implementation.

**Comment by MutatePat** [comment]
I believe the best approach is to keep the orginial CSS intact, and perhaps use a separate CSS file to overwrite the colour schema? 

I'm not sure how feasible it is to intercept the CSS file, however I don't believe it is practical due to potential issues and added complexity. 

Caching the generated less.js seems like a simple enough workaround, but only if we can compile the less files during the Blazor boot sequence ( since that itself usually takes 5 seconds without a cache). As this would make the extra processing unnoticeable.

**Comment by Hona** [comment]
@MutatePat I agree that the original CSS file should stay the same. Only if the middleware is setup to intercept it, then the file response would be overridden.

This solution should make it pretty simple to turn on a dark theme, without changing the _Host.cshtml for example. 

Something like:
```cs
services.AddAntDesign(options =>
{
    options.Theme = OfficialTheme.Dark;
    // options.Theme = OfficialTheme.Compact;
    // options.Theme = OfficialTheme.AliyunConsole
});

...

app.UseAntDesign();
app.UseStaticFiles();
```

In my opinion this is the best solution - user's wanting some pretty components without much configuration will be able to do so, and those who want basic theming (or more custom theming with that options lambda idea) would be able to do so.

**Comment by anddrzejb** [comment]
Coming form a person with very little understanding of theming - all proposals seem very reasonable to me with fair amount of customization capabilities. I read the issues mentioned and it seems to me that some parts of the aforementioned solution(s?) were already tried. @Hona are you willing to implement the proposed solutions?

**Comment by Hona** [comment]
I will implement this when I get some free time - but I'm happy to provide some guidance if someone beats me to it. 

The only concern is that custom themes (other than the `OfficialTheme.***`) will require Node.js installed on the server. It would easily be able to fallback and provide the official themes though.

**Comment by MutatePat** [comment]
> I will implement this when I get some free time - but I'm happy to provide some guidance if someone beats me to it.
> 
> The only concern is that custom themes (other than the `OfficialTheme.***`) will require Node.js installed on the server. It would easily be able to fallback and provide the official themes though.

I am too concerned by node.js - I don't think end .NET Devs will want to install node due to a lack of understanding. Is there any way we can reimplement/port the required functionality to .NET?

**Comment by Hona** [comment]
I did a quick google and it looks like there is a .NET port of less.js called [dotless](https://github.com/dotless/dotless). It looks very promising from the skim through. 

It was last worked on almost a year ago, but if it doesn't work we could always fork and work on it ourselves.

**Comment by ElderJames** [comment]
antd 4.17 has added the CSS variable and we can follow it up.

https://ant.design/components/config-provider/#components-config-provider-demo-theme

**Comment by ElderJames** [comment]
We now support css variable.
